modules/aws_asg/check-docker-health.bash

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bash
+set -o errexit
+set -o pipefail
+
+# Sometimes, the docker service will be running, but certain commands (docker ps) will hang indefinitely.
+# This script detects this behavior and implodes the instance when it occurs.
+
+main() {
+  local warmup_grace_period=600
+  local post_sleep="${POST_SHUTDOWN_SLEEP}"
+  local sleep_time="${DOCKER_PS_SLEEP_TIME}"
+  local run_d="${RUNDIR}"
+  : "${post_sleep:=300}"
+  : "${sleep_time:=3}"
+  : "${run_d:=/var/tmp/travis-run.d}"
+
+  if [[ -f "${run_d}/implode.confirm" ]]; then
+    __handle_implode_confirm "${run_d}" "${post_sleep}"
+    __die imploded 42
+  fi
+
+  # Don't run this unless instance has been up at least $warmup_grace_period seconds
+  uptime_as_int=$(printf "%.0f\n" "$(awk '{ print $1}' /proc/uptime)")
+  if [[ "${uptime_as_int}" -lt "${warmup_grace_period}" ]]; then
+    echo "Not checking docker health yet, as uptime is still less than ${warmup_grace_period} seconds"
+    __die noop 0
+  fi
+
+  echo "Checking docker health..."
+  result=$(timeout "${sleep_time}"s docker ps) || true
+
+  if [ -z "${result}" ]; then
+    __handle_unresponsive_docker "${run_d}"
+    __die imploding 86
+  fi
+  __die noop 0
+}
+
+__handle_implode_confirm() {
+  local run_d="${1}"
+  local post_sleep="${2}"
+
+  local reason
+  reason="$(cat "${run_d}/implode.confirm" 2>/dev/null)"
+  : "${reason:=not sure why}"
+  echo "${SHUTDOWN:-/sbin/shutdown}" -P now "imploding because ${reason}"
+  sleep "${post_sleep}"
+}
+
+__handle_unresponsive_docker() {
+  local run_d="${1}"
+  echo "docker appears to be unhealthy" |
+    tee "${run_d}/implode"
+  "${DOCKER:-docker}" kill -s INT travis-worker
```

Comment:
If docker is unresponsive, won't this `docker kill` block?  If the instance is being imploded, I think this line can be safely removed 😺

Reply:
Touche, removed